Update Storage Access API content (#26558)
mdn/content: files/en-us/web/api/document/hasstorageaccess/index.md

{{APIRef("Storage Access API")}}

> The **`hasStorageAccess()`** method of the {{domxref("Document")}} interface returns a {{jsxref("Promise")}} that resolves with a boolean value indicating whether the document has access to its first-party cookies.

Author: I thought quite a lot about this and ended up deciding to just say "access to cookies" — what we are talking about is the default behavior of cookies, right? So I don't think we do need to give it some kind of fancy name or label — to do so will just end up confusing people, IMO. I've made these changes across all relevant articles. 

I have used the term "partitioned" cookies when referring to cookies given their own separate cookie jar via CHIPs or TCP. I've also added a note to the main API landing page that makes it clear that partitioned cookies don't need the Storage Access API.

Let me know what you think.

> The **`hasStorageAccess()`** method of the {{domxref("Document")}} interface returns a {{jsxref("Promise")}} that resolves with a boolean value indicating whether the document has access to its first-party cookies.

Reviewer: The problem with just saying "cookies" is that it's confusing when it comes to Firefox, since (IIUC) cookies are *always* available, regardless of whether the Storage Access API has been used or not. And same in Chrome: now that CHIPS is available, sites can always use cookies (even when they're in a cross-site iframe) without using the Storage Access API first -- it's just that the cookies have to have the `Partitioned` attribute.

I worry that just saying "cookies" will lead people to think they can't use `Partitioned` cookies until they go through the Storage Access API workflow, or can't use cookies at all on Firefox until they go through `document.requestStorageAccess()`.

> The **`hasStorageAccess()`** method of the {{domxref("Document")}} interface returns a {{jsxref("Promise")}} that resolves with a boolean value indicating whether the document has access to its first-party cookies.

Author: Ah, I get your concern. I've updated all pages to say "unpartitioned cookies", and I've also added a note to the main Storage Access API page to explain what we mean by unpartitioned and partitioned cookies.

> The **`hasStorageAccess()`** method of the {{domxref("Document")}} interface returns a {{jsxref("Promise")}} that resolves with a boolean value indicating whether the document has access to its first-party cookies.

Reviewer: This looks great, thanks!

### Return value

A {{jsxref("Promise")}} that resolves with a boolean value indicating whether the document has access to unpartitioned cookies — `true` if it does, and `false` if not.

The result returned by this method can be inaccurate in a couple of circumstances:

> 1. The user may have browser settings active that block unpartitioned cookies entirely — in this case `true` may be returned even though unpartitioned cookies are still inaccessible. To handle such a situation, you are advised to gracefully handle any errors resulting in cookie values being unretrievable, for example by telling the user that unfortunately access to their personalized settings was blocked, and inviting them to sign in again to make use of them.

Reviewer 2: minor reword
```suggestion
1. The user may have browser settings active that block unpartitioned cookies entirely — in this case `true` may be returned even though unpartitioned cookies are still inaccessible. To handle such a situation, you should gracefully handle any errors resulting in cookie values being unretrievable; for example, inform the user that access to their personalized settings is blocked and invite them to sign in again to use them.
```

> 2. The browser might not block third-party cookies by default — in this case `false` may be returned even though first-party cookies are accessible, and storage access wouldn't need to be requested (i.e. via {{domxref("Document.requestStorageAccess()")}}). This situation will probably only arise in older browsers that don't support `hasStorageAccess()` anyway, so can be handled using feature detection.

Author: Right, of course, this is still the default behavior for most browsers! I've updated this line to say that "This situation is not too hard to handle, as a `Document.requestStorageAccess()` call will resolve successfully when third-party cookies are not blocked, and your code can continue to run." Does that sound OK?

> 2. The browser might not block third-party cookies by default — in this case `false` may be returned even though first-party cookies are accessible, and storage access wouldn't need to be requested (i.e. via {{domxref("Document.requestStorageAccess()")}}). This situation will probably only arise in older browsers that don't support `hasStorageAccess()` anyway, so can be handled using feature detection.

Reviewer: > as a Document.requestStorageAccess() call will resolve successfully when third-party cookies are not blocked

Strictly speaking, that's not guaranteed by the Storage Access API spec. (It's likely that that's what Chrome will do, I think, but I can't make any statements about other browsers' behavior.) So I don't think we should claim the API will have that behavior, generally.

I would encourage developers to check whether cookies are accessible (almost like feature detection, e.g. via `document.cookie`) in order to decide whether they need to call `document.requestStorageAccess()`. That has the highest likelihood of working in any browser, IMO.

> 2. The browser might not block third-party cookies by default — in this case `false` may be returned even though first-party cookies are accessible, and storage access wouldn't need to be requested (i.e. via {{domxref("Document.requestStorageAccess()")}}). This situation will probably only arise in older browsers that don't support `hasStorageAccess()` anyway, so can be handled using feature detection.

Author: OK, I've replaced that bit with "To get around this issue, you could query `Document.cookie` to find out whether your cookies are accessible, and call `Document.requestStorageAccess()` if they are not."

});
```

> > **Note:** See [Using the Storage Access API](/docs/Web/API/Storage_Access_API/Using) for a more complete example.

Reviewer 2: ```suggestion
> **Note:** See [Using the Storage Access API](/en-US/docs/Web/API/Storage_Access_API/Using) for a more complete example.
```
